### PR TITLE
pre-serialize attributes_for_mixer_proxy

### PR DESCRIPTION
**Pull request by kyessenov**
Also fixes the stale pointer dereference.

### CODE REVIEW

**Comment by kyessenov** [comment]
/hold
still segfaulting...

**Comment by kyessenov** [comment]
/hold cancel

confirmed this fixes the segfault

File: src/envoy/http/mixer/control.cc [hunk 1]
```diff
@@ -45,9 +45,21 @@ Control::Control(const Config& config, Upstream::ClusterManager& cm,
...
+  if (!config_.config_pb()
+           .transport()
+           .attributes_for_mixer_proxy()
+           .attributes()
+           .empty()) {
+    config_.config_pb()
+        .transport()
+        .attributes_for_mixer_proxy()
+        .SerializeToString(&serialized_forward_attributes);
+  }
```
Comment (qiwzhang) [inline src/envoy/http/mixer/control.cc]: This is bad for performance.  It is done per request, per Check call.  Could you do following
1) similar to check_client_factory and report_client_factory define here https://github.com/istio/proxy/blob/master/src/envoy/http/mixer/control.h#L54

add the serialized_forward_attributes string at Control class.

2) then just pass its reference to all other paces.

Comment (kyessenov) [reply src/envoy/http/mixer/control.cc]: Done

File: src/envoy/utils/mixer_control.h [hunk 2]
```diff
@@ -31,8 +31,8 @@ void CreateEnvironment(Event::Dispatcher &dispatcher,
                        Runtime::RandomGenerator &random,
                        Grpc::AsyncClientFactory &check_client_factory,
                        Grpc::AsyncClientFactory &report_client_factory,
-                       ::istio::mixerclient::Environment *env,
-                       const Attributes &forward_attributes);
+                       std::string serialized_forward_attributes,
```
Comment (qiwzhang) [inline src/envoy/utils/mixer_control.h]: please use const &

File: src/envoy/utils/grpc_transport.h [hunk 3]
```diff
@@ -62,7 +62,7 @@ class GrpcTransport : public Grpc::TypedAsyncRequestCallbacks<ResponseType>,
 
   Grpc::AsyncClientPtr async_client_;
   ResponseType* response_;
-  const ::istio::mixer::v1::Attributes& forward_attributes_;
+  std::string serialized_forward_attributes_;
```
Comment (qiwzhang) [inline src/envoy/utils/grpc_transport.h]: I think you can use const &,  not need to copy the data

File: src/envoy/http/mixer/control.cc [hunk 4]
```diff
@@ -33,21 +33,30 @@ Control::Control(const Config& config, Upstream::ClusterManager& cm,
...
+  if (!config_.config_pb()
+           .transport()
+           .attributes_for_mixer_proxy()
+           .attributes()
+           .empty()) {
+    config_.config_pb()
+        .transport()
+        .attributes_for_mixer_proxy()
+        .SerializeToString(&serialized_forward_attributes_);
+  }
```
Comment (qiwzhang) [inline src/envoy/http/mixer/control.cc]: Can we define a function to share this code with tcp?   similar to Utils::CreateEnvironment()

File: src/envoy/utils/mixer_control.h (hunk 2 above)
Comment (kyessenov) [reply src/envoy/utils/mixer_control.h]: TCP control doesn't own the string though? Do you want me to store it there?

Comment (qiwzhang) [reply src/envoy/utils/mixer_control.h]: Yes.

File: src/envoy/http/mixer/control.cc (hunk 4 above)
Comment (kyessenov) [reply src/envoy/http/mixer/control.cc]: Done

File: src/envoy/utils/grpc_transport.h (hunk 3 above)
Comment (kyessenov) [reply src/envoy/utils/grpc_transport.h]: Done

File: src/envoy/utils/mixer_control.h (hunk 2 above)
Comment (kyessenov) [reply src/envoy/utils/mixer_control.h]: Done

**Pull request merged**